src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs

```diff
@@ -78,10 +86,55 @@ private static string UrlEncodeSegment(byte[] segment)
             return Base64UrlHelpers.Encode(segment);
         }
 
-        private static string EncodeHeaderToJson(X509Certificate2 certificate, string base64EncodedThumbprint, bool sendX5C)
+        private static string EncodeHeaderToJson(X509Certificate2 certificate, bool sendX5C, bool useSha2AndPss)
+        {
+            string thumbprint = ComputeCertThumbprint(certificate, useSha2AndPss);
+
+            string alg = useSha2AndPss ? "PS256" : "RS256";
+            string thumbprintKey = useSha2AndPss ? "x5t#S256" : "x5t";
+            string header;
+
+            if (sendX5C)
+            {
+#if NETFRAMEWORK
+                string x5cValue = Convert.ToBase64String(certificate.GetRawCertData());
+#else
+                string x5cValue = Convert.ToBase64String(certificate.RawData);
+#endif
+                header = $$"""{"alg":"{{alg}}","typ":"JWT","{{thumbprintKey}}":"{{thumbprint}}","x5c":"{{x5cValue}}"}""";
```

Comment:
This is much faster than creating an Object with these properties and applying JSON serialization.